### PR TITLE
[framework] functional tests are now run per folder to prevent memory limits

### PR DESCRIPTION
**Pull request by TomasLudvik**
#### Description, the reason for the PR
There is problem with functional tests, that with growing number of them the memory is increasing and tests are slowing down when run in single phpunit suite. This change introduces running phpunit tests per folder that reduces memory and increases speed of tests by up to 60%. This has been inspired by @RostislavKreisinger 

#### Fixes issues
... 

#### License Agreement for contributions

- [ ] I have read and signed [License Agreement for contributions](https://www.shopsys.com/license-agreement)

----
:globe_with_meridians: Live Preview:
  - https://tl-run-functional-tests-per-folder.odin.shopsys.cloud
  - https://cz.tl-run-functional-tests-per-folder.odin.shopsys.cloud

### CODE REVIEW

File: packages/framework/build.xml [hunk 1]
```diff
@@ -141,12 +146,14 @@
...
 
-    <target name="build-dev-quick" depends="build-version-generate,clean,composer-dev,dirs-create,test-dirs-create,assets,npm,db-migrations,domains-data-create,friendly-urls-generate,friendly-url-entity-mapping-check,domains-urls-replace,elasticsearch-index-migrate,clean-redis-old" description="Builds application for development preserving your DB while skipping nonessential steps."/>    <target name="build-dev-quick" depends="build-version-generate,clean,composer-dev,dirs-create,test-dirs-create,assets,npm,db-migrations,domains-data-create,friendly-urls-generate,friendly-url-entity-mapping-check,domains-urls-replace,elasticsearch-index-migrate,clean-redis-old" description="Builds application for development preserving your DB while skipping nonessential ."/>
+    <target name="build-dev-quick" depends="build-version-generate,clean,composer-dev,dirs-create,test-dirs-create,assets,npm,db-migrations,domains-data-create,friendly-urls-generate,friendly-url-entity-mapping-check,domains-urls-replace,elasticsearch-index-migrate,clean-redis-old" description="Builds application for development preserving your DB while skipping nonessential ."/>
```
Comment (vitek-rostislav) [inline packages/framework/build.xml]: was it duplicated? Lol :smile: However, it is still duplicated now - you are adding the `build-dev-quick` target again few lines below. Let's keep just one :slightly_smiling_face:

File: packages/framework/build.xml [hunk 2]
```diff
@@ -1414,42 +1421,117 @@
...
+            <then>
+                <phingcall target="tests-set-failed"/>
+            </then>
+        </if>
     </target>
 
     <target name="tests-functional-frontend-api" depends="production-protection,clean,test-clean-redis,domains-info-load" description="Runs Frontend API functional tests." hidden="true">
-        <exec executable="${path.phpunit.executable}" logoutput="true" passthru="true" checkreturn="true">
+        <phingcall target="tests-reset-failed"/>
+        <foreach param="testsDirname" target="tests-functional-frontend-api-run" inheritrefs="true" inheritall="true">
```
Comment (vitek-rostislav) [inline packages/framework/build.xml]: are the `inheritrefs` and `inheritall` settings necessary? You do not define them in the `tests-functional-application` above :thinking:

**Review by vitek-rostislav — CHANGES_REQUESTED** [review]
My notes and feelings from testing:
1. `tests-functional-application`
**1m  12.73s on this branch vs 1m 16.22s on `16.0`** - that is probably because of the fact that most of the functional tests are located in `/Model` folder that is still run at once, if we wanted to speed the tests even more, we should split the runs into the subfolders.
2. `tests-functional-frontend-api`
**3m 16.56s on this branch vs 7m 17.35s on `16.0`** - now THAT is awesome! :heart_eyes: 

When a test fails, it might be now confusing a little bit as there is quite a lot of information, I see the following error which might not be clear enough for the developers that are used to the regular PHPUnit output we used until now:
```
Shopsys Platform > tests-reset-failed:

   [delete] Deleting: /var/www/html/project-base/app/var/PHING_TESTS_FAILED
[phingcall] [Phing\Exception\BuildException] Tests failed

BUILD FAILED
/var/www/html/packages/framework/build.xml:1447:47 Execution of the target buildfile failed. Aborting.

Total time: 1 minute  14.14 seconds
```

I need to scroll up in the output to find that important information:
image

None of these are blockers essentially, I believe we can live with that, the speed up of the API tests is probably worth the downsides of the pull request. However, I still believe the best solution would be should be figuring out and fixing the memory leak itself.

**Review by vitek-rostislav — APPROVED** [review]

**Review by pk16011990 — APPROVED** [review]

**Comment by henzigo** [comment]
Shouldn't this issue be closed #3196 ?

**Pull request merged**